src/vmm/src/vmm_config/drive.rs

```diff
@@ -70,109 +81,114 @@ pub struct BlockDeviceConfig {
-        // check if the path exists
-        if !block_device_config.path_on_host.exists() {
-            return Err(DriveError::InvalidBlockDevicePath);
-        }
-
-        if self
-            .get_index_of_drive_path(&block_device_config.path_on_host)
-            .is_some()
-        {
-            return Err(DriveError::BlockDevicePathAlreadyExists);
-        }
-
-        // check whether the Device Config belongs to a root device
-        // we need to satisfy the condition by which a VMM can only have on root device
-        if block_device_config.is_root_device {
-            if self.has_root_block {
-                return Err(DriveError::RootBlockDeviceAlreadyAdded);
-            } else {
-                self.has_root_block = true;
-                // Root Device should be the first in the list whether or not PARTUUID is specified
-                // in order to avoid bugs in case of switching from partuuid boot scenarios to
-                // /dev/vda boot type.
-                self.config_list.push_front(block_device_config);
+    pub fn insert(&mut self, config: BlockDeviceConfig) -> Result<()> {
+        let is_root_device = config.is_root_device;
+        let position = self.get_index_of_drive_id(&config.drive_id);
+        let block_dev = Arc::new(Mutex::new(Self::create_block(config)?));
+        // If the id of the drive already exists in the list, the operation is update/overwrite.
+        match position {
+            // New block device.
+            None => {
+                // Check whether the Device Config belongs to a root device,
+                // we need to satisfy the condition by which a VMM can only have one root device.
+                if is_root_device {
+                    if self.has_root_block {
+                        return Err(DriveError::RootBlockDeviceAlreadyAdded);
+                    } else {
+                        // Root Device should be the first in the list whether or not PARTUUID is
+                        // specified in order to avoid bugs in case of switching from partuuid boot
+                        // scenarios to /dev/vda boot type.
+                        self.list.push_front(block_dev);
+                        self.has_root_block = true;
+                    }
+                } else {
+                    self.list.push_back(block_dev);
+                }
+            }
+            // Update existing block device.
+            Some(mut index) => {
+                // Check if the root block device is being updated.
+                if index == 0 && self.has_root_block {
+                    // Set root flag according to the updated block config.
+                    self.has_root_block = is_root_device;
+                } else if is_root_device {
+                    // Check if a second root block device is being added.
+                    if self.has_root_block {
+                        return Err(DriveError::RootBlockDeviceAlreadyAdded);
```

Comment:
Checking for duplicate block device at top will greatly simplify this code.

Reply:
Right, will do.